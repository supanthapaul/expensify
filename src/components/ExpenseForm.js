import React from 'react';
//http://momentjs.com
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {

   constructor(props) {
      super(props);

      this.state = {
         description: props.expense ? props.expense.description : '',
         note: props.expense ? props.expense.note : '',
         amount: props.expense ? (props.expense.amount / 100).toString() : '',   // convert to dollar
         createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
         calenderFocused: false,
         error: ''
      };
   }

   onDescriptionChange = (e) => {
      const description = e.target.value;
      this.setState(() => ({
         description
      }));
   };
   onNoteChange = (e) => {
      const note = e.target.value;
      this.setState(() => ({
         note
      }))
   };
   onAmountChange = (e) => {
      const amount = e.target.value;
      // either blank amount or formatted amount
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState(() => ({
            amount
         }));
      }
   };
   onDateChange = (createdAt) => {
      if (createdAt) {
         this.setState(() => ({
            createdAt
         }));
      }
   };
   onFocusChange = ({ focused }) => {
      this.setState(() => ({
         calenderFocused: focused
      }))
   };

   onSubmit = (e) => {
      e.preventDefault();

      if (!this.state.description || !this.state.amount) {
         // set error state
         this.setState(() => ({ error: 'Please provide Description and Amount.' }));
      } else {
         // clear the error
         this.setState(() => ({ error: '' }));

         // onSubmit need to be passed whereever this component is used
         // form inputs will be passed to the function
         this.props.onSubmit({
            description: this.state.description,
            amount: parseFloat(this.state.amount, 10) * 100, // format to cents
            createdAt: this.state.createdAt.valueOf(),   // get the timestamp from date picker
            note: this.state.note
         });
      }
   }

   render() {
      return (
         <form className="form" onSubmit={this.onSubmit}>
            {this.state.error && <p className="form__error">{this.state.error}</p>}
            <input
               type="text"
               className="text-input"
               placeholder="Description"
               autoFocus
               value={this.state.description}
               onChange={this.onDescriptionChange}
            />
            <input
               type="text"
               className="text-input"
               placeholder="Amount"
               value={this.state.amount}
               onChange={this.onAmountChange}
            />
            {/* Using react-dates library */}
            {/* https://github.com/airbnb/react-dates */}
            <SingleDatePicker
               date={this.state.createdAt}
               onDateChange={this.onDateChange}
               focused={this.state.calenderFocused}
               onFocusChange={this.onFocusChange}
               numberOfMonths={1}
               // Make every date available
               isOutsideRange={() => false}
            />
            <textarea
               placeholder="Add a note for your expense (optional)"
               className="text-area"
               value={this.state.note}
               onChange={this.onNoteChange}
            >
            </textarea>
            <div>
               <button className="button">Save Expense</button>
            </div>
         </form>
      );
   }
}

export default ExpenseForm;