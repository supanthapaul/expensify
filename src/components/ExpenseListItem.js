import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>${amount} - {createdAt} </p>
    </div>
);

// here we don't need any values from the store
// we just connect the component to the store
// so that we get access to dispatch
export default ExpenseListItem;