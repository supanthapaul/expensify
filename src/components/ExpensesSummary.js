import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
   const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
   const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
   return (
      <h1>Viewing {expensesCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
   )
}

const mapStateToProps = (state) => {
   const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
   return {
      expensesCount: visibleExpenses.length,
      expensesTotal: getExpensesTotal(visibleExpenses)
   }
}

export default connect(mapStateToProps)(ExpensesSummary);