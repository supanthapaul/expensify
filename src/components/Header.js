import React from 'react'
import { NavLink } from 'react-router-dom'

const Header = () => (
    <div>
        <h1>Expensify</h1>
        <ul>
            <li><NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink></li>
            <li><NavLink to="/create" activeClassName="is-active">Create Expense</NavLink></li>
            <li><NavLink to="/help" activeClassName="is-active">Help</NavLink></li>
        </ul>
    </div>
);

export default Header;