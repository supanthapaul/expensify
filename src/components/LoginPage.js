import React from 'react'
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => {
   return (
      <div>
         <h1>Welcome To Expensify</h1>
         <button onClick={startLogin}>Login with Google</button>
      </div>
   )
}

const mapDispatchToProps = (dispatch) => ({
   startLogin: () => dispatch(startLogin())
});

export default connect(null, mapDispatchToProps)(LoginPage);