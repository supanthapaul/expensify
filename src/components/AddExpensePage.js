import React, { Component } from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends Component {

   onSubmit = (expense) => {
      // handle submit
      this.props.addExpense(expense);
      // redirect to dashboard
      this.props.history.push('/');
   }

   render() {
      return (
         <div>
            <p>Add Expense</p>
            <ExpenseForm
               onSubmit={this.onSubmit}
            />
         </div>
      )
   }
}

const mapDispatchToProps = (dispatch) => ({
   addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(null, mapDispatchToProps)(AddExpensePage);