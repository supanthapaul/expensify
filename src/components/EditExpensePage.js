import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses'


export class EditExpensePage extends Component {

   onSubmit = (expense) => {
      // dispatch the action to edit the expense
      this.props.startEditExpense(this.props.expense.id, expense);
      // redirect to dashboard
      this.props.history.push('/');
   }
   onRemoveClick = () => {
      // dispatch the action to remove the expense
      this.props.startRemoveExpense({ id: this.props.expense.id });
      // redirect to dashboard
      this.props.history.push('/');
   }

   render() {
      return (
         <div>
            <div className="page-header">
               <div className="content-container">
                  <h1 className="page-header__title">Edit Expense</h1>
               </div>
            </div>
            <div className="content-container">
               <ExpenseForm
                  expense={this.props.expense}
                  onSubmit={this.onSubmit}
               />
               <button className="button button--secondary" onClick={this.onRemoveClick}
               >Remove Expense</button>
            </div>
         </div>
      )
   }
}

const mapStateToProps = (state, props) => {
   return {
      expense: state.expenses.find((expense) => expense.id === props.match.params.id)
   }
}

const mapDispatchToProps = (dispatch) => ({
   startEditExpense: (expenseId, updates) => dispatch(startEditExpense(expenseId, updates)),
   startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);