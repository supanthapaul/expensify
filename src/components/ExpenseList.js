import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

// export unconnected component for testing
export const ExpenseList = (props) => (
   <div className="content-container">
      <div className="list-header">
         <div className="show-for-mobile">Expenses</div>
         <div className="show-for-desktop">Expense</div>
         <div className="show-for-desktop">Amount</div>
      </div>
      <div className="list-body">
         {
            props.expenses.length === 0 ? (
               <div className="list-item list-item--message">
                  <span>No expenses.</span>
               </div>
            ) : (
                  props.expenses.map((expense) => {
                     return <ExpenseListItem key={expense.id} {...expense} />
                  })
               )
         }
      </div>

   </div>
);

const mapStateToProps = (state) => {
   // return object with the things you wanna grab from the store
   // it gets passed into the component as props
   return {
      expenses: selectExpenses(state.expenses, state.filters)
   }
};

// create and export the higher order component
export default connect(mapStateToProps)(ExpenseList);
