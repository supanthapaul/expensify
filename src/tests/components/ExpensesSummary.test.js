import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should render ExpensesSummary correctly with 1 expense', () => {
   const wrapper = shallow(<ExpensesSummary
      expensesCount={1}
      expensesTotal={195} />);

   expect(wrapper).toMatchSnapshot();
});

test('Should render ExpensesSummary correctly with multiple expenses', () => {
   const wrapper = shallow(<ExpensesSummary
      expensesCount={2}
      expensesTotal={12132323} />);

   expect(wrapper).toMatchSnapshot();
});