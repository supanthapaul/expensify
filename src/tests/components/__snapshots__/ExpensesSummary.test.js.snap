// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary correctly with 1 expense 1`] = `
<h1>
  Viewing 
  1
   
  expense
   totalling 
  $1.95
</h1>
`;

exports[`Should render ExpensesSummary correctly with multiple expenses 1`] = `
<h1>
  Viewing 
  2
   
  expenses
   totalling 
  $121,323.23
</h1>
`;
